File: internal/pricing/pricing.go
// Package pricing is the one price source for the cross-harness cost view:
// per-(provider, model) USD per 1M tokens from models.dev.
//
// Two sources, one lookup. The embedded snapshot (models.json, refreshed by
// go generate from the committed models.dev fixture) always works offline.
// The cache file ($XDG_CACHE_HOME/krowk/models.json, refreshed only by an
// explicit refresh — `krowk pricing refresh`, and later `krowk sessions sync`)
// wins when present and parseable, so newer prices apply without a release.
// Lookup order is cache first, embedded second; a corrupt cache falls back to
// embedded without an error.
//
// Price never touches the network: the hot paths (sessions list/show, import,
// doctor) call it per row and must work offline and fast. The snapshot parses
// once per process and the cache file re-parses only when its path or mtime
// changes, so a listing of ten thousand rows parses nothing per row. Refresh
// is the only function that dials out, and it takes its HTTP client as an
// argument — tests hand it a failing transport to prove the hot paths hold.
//
// The environment is injected, never read: Bind sets the lookup the package
// uses (the CLI binds the process environment once at startup), CachePath and
// Refresh take an Env directly, and nothing here reads the process environment. The
// models.dev URL lives in this package and nowhere else.
//
// Prices are provider-specific — Bedrock's us.anthropic.claude-fable-5-1 is
// not Anthropic's claude-fable-5-1 — so every lookup is keyed (provider,
// model), never model alone. Each importer normalises its own model string to
// the models.dev id for its provider before calling Price (see Normalize);
// an unknown pair answers ok=false and the caller renders "—", never 0.
//
// Rates are current, never historical: sessions show footnotes "priced at
// current models.dev rates (snapshot YYYY-MM-DD)" from SnapshotDate.
package pricing

//go:generate go run ./generate -date 2026-09-10

import (
	_ "embed"
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// ModelsURL is where the full price file lives. Only this package names it:
// Refresh fetches it, the generate script can fetch it behind -live, and no
// other package may dial it for prices.
const ModelsURL = "https://models.dev/api.json"

//go:embed models.json
var embeddedJSON []byte

// Env is a lookup function, the same shape internal/store, internal/harness
// and internal/runctx use, so a test moves the cache by handing over a
// different environment instead of touching the process one.
type Env func(string) string

// Rates are USD per 1M tokens for one (provider, model) pair. Reasoning is
// models.dev's reasoning rate when the model publishes one; HasReasoning says
// whether it did, because a published 0 and no published rate price
// differently — without one, reasoning tokens fall back to the output rate.
type Rates struct {
	Input        float64 `json:"input"`
	Output       float64 `json:"output"`
	CacheRead    float64 `json:"cache_read"`
	CacheWrite   float64 `json:"cache_write"`
	Reasoning    float64 `json:"reasoning,omitempty"`
	HasReasoning bool    `json:"has_reasoning,omitempty"`
}

// Tokens are the token counts a cost is derived from.
type Tokens struct {
	Input      int64
	Output     int64
	CacheRead  int64
	CacheWrite int64
	Reasoning  int64
}

// Cost derives the USD price: the sum over kinds of tokens/1e6 × rate.
// Reasoning tokens use the reasoning rate when the model publishes one, else
// the output rate. Nothing about the result is persisted: message.cost_usd
// stays source-reported only, and this is computed at display time.
func (r Rates) Cost(t Tokens) float64 {
	const perMillion = 1e6
	reasoning := r.Reasoning
	if !r.HasReasoning {
		reasoning = r.Output
	}
	return float64(t.Input)/perMillion*r.Input +
		float64(t.Output)/perMillion*r.Output +
		float64(t.CacheRead)/perMillion*r.CacheRead +
		float64(t.CacheWrite)/perMillion*r.CacheWrite +
		float64(t.Reasoning)/perMillion*reasoning
}

// Normalize maps an importer's model string to the models.dev id for its
// provider before lookup. Today it is the identity: Claude JSONL already
// names claude-fable-5-1 under anthropic and Bedrock ARNs arrive as
// us.anthropic.claude-fable-5-1 under amazon-bedrock, both of which are the
// models.dev ids. It exists so the per-importer mappings the plan calls for
// have one seam to grow in when an importer disagrees, rather than each
// call site string-munging on its own.
func Normalize(provider, model string) string {
	return model
}

type key struct {
	provider string
	model    string
}

var (
	mu        sync.Mutex
	bound     Env
	embedded  map[key]Rates
	embedErr  error
	embedOnce sync.Once

	cachePath string
	cacheMod  time.Time
	cacheSize int64
	cached    map[key]Rates
	cacheSeen bool
)

// Bind sets the environment Price resolves the cache file against. The CLI
// calls it once at startup with the process environment; tests bind a fake.
// Unbound, Price falls back to the platform cache dir. Either way the lookup
// reads files only — never the network, never the process environment.
func Bind(env Env) {
	mu.Lock()
	defer mu.Unlock()
	bound = env
	cacheSeen = false
	cached = nil
}

// CachePath is where the refreshed price file lives:
// $XDG_CACHE_HOME/krowk/models.json, falling back to ~/.cache/krowk/.
// XDG_CACHE_HOME must be absolute to count (the XDG basedir rule); a missing
// home is an empty string, never a guess — a cache that invented a home
// would write prices onto a machine nobody is using.
func CachePath(env Env) string {
	if env != nil {
		if dir := env("XDG_CACHE_HOME"); filepath.IsAbs(dir) {
			return filepath.Join(dir, "krowk", "models.json")
		}
		if home := env("HOME"); filepath.IsAbs(home) {
			return filepath.Join(home, ".cache", "krowk", "models.json")
		}
		return ""
	}
	if dir, err := os.UserCacheDir(); err == nil && dir != "" {
		return filepath.Join(dir, "krowk", "models.json")
	}
	return ""
}

// Price answers the rates for one (provider, model) pair: the cache file
// when present and parseable, else the embedded snapshot. Unknown pairs
// answer ok=false — the caller renders "—", never 0.
func Price(provider, model string) (Rates, bool) {
	embedOnce.Do(func() {
		embedded, embedErr = parseRates(embeddedJSON)
	})
	mu.Lock()
	defer mu.Unlock()
	if embedErr == nil {
		if r, ok := lookupLocked(provider, model); ok {
			return r, true
		}
		if r, ok := embedded[key{provider, model}]; ok {
			return r, true
		}
	}
	return Rates{}, false
}

// lookupLocked consults the cache file, reloading it only when its path,
// mtime or size moved since the last load. A missing file is not an error —
// it means no refresh has run yet. A corrupt one falls back to embedded
// without an error, so a half-written refresh can never break pricing.
func lookupLocked(provider, model string) (Rates, bool) {
	path := CachePath(bound)
	if path == "" {
		return Rates{}, false
	}
	info, err := os.Stat(path)
	if err != nil {
		cacheSeen = false
		cached = nil
		return Rates{}, false
	}
	if !cacheSeen || cachePath != path || !info.ModTime().Equal(cacheMod) || info.Size() != cacheSize {
		raw, err := os.ReadFile(path)
		if err != nil {
			cacheSeen = false
			cached = nil
			return Rates{}, false
		}
		parsed, err := parseRates(raw)
		if err != nil {
			cacheSeen = false
			cached = nil
			return Rates{}, false
		}
		cached = parsed
		cachePath = path
		cacheMod = info.ModTime()
		cacheSize = info.Size()
		cacheSeen = true
	}
	r, ok := cached[key{provider, model}]
	return r, ok
}

// parseRates reads either shape prices come in: the trimmed snapshot
// ({provider: {model: cost}}) and the full models.dev file
// ({provider: {models: {id: {cost...}}, ...}}). The discriminator is the
// "models" key full-shape providers carry and snapshot providers never do.
// Only the numeric base per-token rates are kept — tiered overrides and
// modality rates are skipped value by value, so a refreshed full-file cache
// with tomorrow's models still parses. Extra providers and models pass
// through: the generate script is what trims the snapshot, not this.
func parseRates(raw []byte) (map[key]Rates, error) {
	var top map[string]map[string]json.RawMessage
	if err := json.Unmarshal(raw, &top); err != nil {
		return nil, err
	}
	out := make(map[key]Rates)
	for provider, fields := range top {
		modelsRaw, ok := fields["models"]
		if !ok {
			for model, costRaw := range fields {
				var cost map[string]json.RawMessage
				if err := json.Unmarshal(costRaw, &cost); err != nil {
					continue
				}
				if r, ok := ratesFromRaw(cost); ok {
					out[key{provider, model}] = r
				}
			}
			continue
		}
		var models map[string]struct {
			Cost map[string]json.RawMessage `json:"cost"`
		}
		if err := json.Unmarshal(modelsRaw, &models); err != nil {
			continue
		}
		for model, m := range models {
			if r, ok := ratesFromRaw(m.Cost); ok {
				out[key{provider, model}] = r
			}
		}
	}
	return out, nil
}

// ratesFromRaw keeps the numeric base rates out of a cost object and reports
// whether it found any. Non-numeric members (tiers, context_over_200k) are
// skipped, and a cost with no numeric rate at all is not a price.
func ratesFromRaw(cost map[string]json.RawMessage) (Rates, bool) {
	var r Rates
	found := false
	take := func(name string, dst *float64) {
		raw, ok := cost[name]
		if !ok {
			return
		}
		var v float64
		if err := json.Unmarshal(raw, &v); err != nil {
			return
		}
		*dst, found = v, true
	}
	take("input", &r.Input)
	take("output", &r.Output)
	take("cache_read", &r.CacheRead)
	take("cache_write", &r.CacheWrite)
	if raw, ok := cost["reasoning"]; ok {
		var v float64
		if err := json.Unmarshal(raw, &v); err == nil {
			r.Reasoning, r.HasReasoning, found = v, true, true
		}
	}
	return r, found
}

File: crates/krowk/src/pricing.rs
//! Model prices, per million tokens, from models.dev. Two sources, one
//! lookup: the snapshot embedded at build time always works offline, and the
//! cache `pricing refresh` writes is preferred when it holds the pair. Tokens
//! are stored per turn and priced at read time, so a refresh reprices history.
//!
//! Freshness has one recurring path and no machinery: `sessions sync`
//! refreshes the cache when its last fetch is a day old, and `doctor` and
//! `pricing refresh` say how old it is. Nothing on a read path — list, show,
//! import, doctor — touches the network, and there is no timer: prices move
//! monthly, and a scheduler would outlive its value.
//!
//! Audio rates (`input_audio`, `output_audio`) are read by nobody, on
//! purpose: Claude, Cursor and opencode transcripts carry no audio tokens,
//! and the store has no column to keep them in. The one place they could
//! arrive is an OpenAI-shaped usage block (`prompt_tokens_details` /
//! `completion_tokens_details.audio_tokens`) in a provider ledger, where
//! they are counted inside input and output and priced at the text rate —
//! an undercount, since audio costs more. Revisit when a ledger row or an
//! importer reports a non-zero audio count: reading the rates is a field
//! here and a column in the store, not a re-fetch — the cache keeps them as
//! models.dev wrote them.

use serde_json::value::RawValue;
use serde_json::{Map, Value};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::Mutex;
use std::time::Duration;

pub const MODELS_URL: &str = "https://models.dev/api.json";
/// The models.dev snapshot the embedded prices were trimmed from.
pub const SNAPSHOT_DATE: &str = "2026-09-10";
const EMBEDDED: &str = include_str!("models.json");

const REFRESH_TIMEOUT: Duration = Duration::from_secs(5);
const MAX_BODY: u64 = 32 << 20;
const MAX_ETAG: usize = 4096;

#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct Rates {
    pub input: f64,
    pub output: f64,
    pub cache_read: f64,
    pub cache_write: f64,
    pub reasoning: f64,
    pub has_reasoning: bool,
}

#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct Tokens {
    pub input: i64,
    pub output: i64,
    pub cache_read: i64,
    pub cache_write: i64,
    pub reasoning: i64,
}

impl Rates {
    /// USD for these tokens. Reasoning is billed at the output rate when the
    /// model publishes none of its own; negative counts cost nothing.
    pub fn cost(&self, t: Tokens) -> f64 {
        let n = |x: i64| x.max(0) as f64 / 1e6;
        let reasoning = if self.has_reasoning { self.reasoning } else { self.output };
        n(t.input) * self.input + n(t.output) * self.output + n(t.cache_read) * self.cache_read + n(t.cache_write) * self.cache_write + n(t.reasoning) * reasoning
    }
}

type Table = HashMap<(String, String), Rates>;

/// Where a price came from, and so how current it is: the refreshed cache as
/// of its last fetch, or the snapshot embedded at build time. Either way the
/// rate is today's, never the one in force when the tokens were spent.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Basis {
    Cache { fetched_at_ms: Option<i64> },
    Snapshot,
}

impl Basis {
    /// The footnote a priced figure carries.
    pub fn note(&self) -> String {
        match self {
            Basis::Cache { fetched_at_ms: Some(ms) } => format!("models.dev prices fetched {}", date_of(*ms)),
            Basis::Cache { fetched_at_ms: None } => "models.dev prices from the refreshed cache".into(),
            Basis::Snapshot => format!("models.dev snapshot {SNAPSHOT_DATE} embedded in this build"),
        }
    }
}

/// Every basis a set of figures was priced on, as one footnote.
pub fn basis_note(bases: &std::collections::BTreeSet<Basis>) -> String {
    let notes: Vec<String> = bases.iter().map(Basis::note).collect();
    if notes.is_empty() {
        return String::new();
    }
    format!("priced at current rates ({}), not the rates in force at the time", notes.join(" and "))
}

fn date_of(ms: i64) -> String {
    jiff::Timestamp::from_millisecond(ms).map(|t| t.strftime("%Y-%m-%d").to_string()).unwrap_or_else(|_| "at an unknown date".into())
}

/// $XDG_CACHE_HOME/krowk/models.json when absolute, else ~/.cache/krowk/.
pub fn cache_path(env: &dyn Fn(&str) -> String) -> Option<PathBuf> {
    let xdg = env("XDG_CACHE_HOME");
    if Path::new(&xdg).is_absolute() {
        return Some(Path::new(&xdg).join("krowk").join("models.json"));
    }
    let home = env("HOME");
    Path::new(&home).is_absolute().then(|| Path::new(&home).join(".cache").join("krowk").join("models.json"))
}

pub fn meta_path(cache: &Path) -> PathBuf {
    cache.with_file_name("models.meta.json")
}

/// The cache (with when it was fetched) and the snapshot, each parsed once
/// per process. They fail independently: a corrupt cache never hides the
/// snapshot.
type Loaded = (Option<(Table, Option<i64>)>, Table);
static LOADED: Mutex<Option<Loaded>> = Mutex::new(None);

/// The rates for a (provider, model) pair: the cache's, else the snapshot's.
pub fn price(env: &dyn Fn(&str) -> String, provider: &str, model: &str) -> Option<Rates> {
    price_with_basis(env, provider, model).map(|(r, _)| r)
}

/// `price`, and which of the two sources answered.
pub fn price_with_basis(env: &dyn Fn(&str) -> String, provider: &str, model: &str) -> Option<(Rates, Basis)> {
    let mut loaded = LOADED.lock().unwrap_or_else(|e| e.into_inner());
    let (cache, embedded) = loaded.get_or_insert_with(|| {
        let cache = cache_path(env).and_then(|p| {
            let table = parse_rates(&std::fs::read(&p).ok()?)?;
            Some((table, fetched_at_ms(&p)))
        });
        (cache, parse_rates(EMBEDDED.as_bytes()).unwrap_or_default())
    });
    let key = (provider.to_string(), model.to_string());
    if let Some((table, fetched)) = cache
        && let Some(r) = table.get(&key)
    {
        return Some((*r, Basis::Cache { fetched_at_ms: *fetched }));
    }
    embedded.get(&key).map(|r| (*r, Basis::Snapshot))
}

/// Either shape prices come in: the trimmed snapshot's
/// `{provider: {model: cost}}`, or models.dev's full
/// `{provider: {models: {model: {cost: …}}}}`. Typed, with every
/// field but `cost` skipped unread: the full file is megabytes, and it is
/// parsed on every `sessions` listing.
fn parse_rates(raw: &[u8]) -> Option<Table> {
    // One bad entry costs that entry, never its provider.
    #[derive(serde::Deserialize)]
    struct FullModel {
        cost: Option<Map<String, Value>>,
    }
    let top: HashMap<String, &RawValue> = serde_json::from_slice(raw).ok()?;
    let mut out = Table::new();
    for (provider, fields) in top {
        let Ok(fields) = serde_json::from_str::<HashMap<String, &RawValue>>(fields.get()) else { continue };
        let full: Vec<(String, Map<String, Value>)> = fields
            .get("models")
            .and_then(|m| serde_json::from_str::<HashMap<String, &RawValue>>(m.get()).ok())
            .map(|models| {
                models
                    .into_iter()
                    .filter_map(|(m, v)| Some((m, serde_json::from_str::<FullModel>(v.get()).ok()?.cost.filter(|c| !c.is_empty())?)))
                    .collect()
            })
            .unwrap_or_default();
        let entries: Vec<(String, Map<String, Value>)> = if full.is_empty() {
            fields.iter().filter_map(|(m, v)| Some((m.clone(), serde_json::from_str::<Map<String, Value>>(v.get()).ok()?))).collect()
        } else {
            full
        };
        for (model, cost) in entries {
            if let Some(r) = rates_from(&cost) {
                out.insert((provider.clone(), model), r);
            }
        }
    }
    Some(out)
}

fn rates_from(cost: &Map<String, Value>) -> Option<Rates> {
    let mut r = Rates::default();
    let mut found = false;
    let mut take = |name: &str, dst: &mut f64| -> bool {
        match cost.get(name).and_then(Value::as_f64).filter(|v| *v >= 0.0) {
            Some(v) => {
                *dst = v;
                found = true;
                true
            }
            None => false,
        }
    };
    take("input", &mut r.input);
    take("output", &mut r.output);
    take("cache_read", &mut r.cache_read);
    take("cache_write", &mut r.cache_write);
    r.has_reasoning = take("reasoning", &mut r.reasoning);
    found.then_some(r)
}

/// What one refresh did.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Outcome {
    /// New prices are in the cache.
    Refreshed,
    /// models.dev confirmed the cache (304): the prices stand, freshly checked.
    Unchanged,
    /// No new prices — no network, no answer, or an answer that is not a
    /// price file. The cache or the snapshot still prices everything.
    Unreachable(String),
}

/// Fetches models.dev into the cache, conditionally on the last ETag. Only a
/// cache that cannot be written (or stamped) is an error.
pub fn refresh(env: &dyn Fn(&str) -> String, url: &str) -> Result<Outcome, String> {
    refresh_within(env, url, REFRESH_TIMEOUT)
}

/// `refresh` bounded by `timeout`: sync's is tighter, since a scheduled sync
/// is not a place to wait on a slow network.
pub fn refresh_within(env: &dyn Fn(&str) -> String, url: &str, timeout: Duration) -> Result<Outcome, String> {
    let path = cache_path(env).ok_or("pricing: no cache directory in environment")?;
    // The ETag is only worth sending for a cache that still holds prices: a
    // 304 against a deleted or corrupt file would confirm prices nobody has.
    let etag = if cache_readable(&path) { load_etag(&path) } else { String::new() };
    let agent: ureq::Agent = ureq::Agent::config_builder()
        .timeout_global(Some(timeout))
        .max_redirects(0)
        .http_status_as_error(false)
        .tls_config(ureq::tls::TlsConfig::builder().root_certs(ureq::tls::RootCerts::PlatformVerifier).build())
        .build()
        .into();
    let mut req = agent.get(if url.is_empty() { MODELS_URL } else { url });
    if !etag.is_empty() {
        req = req.header("If-None-Match", &etag);
    }
    let mut res = match req.call() {
        Ok(res) => res,
        // The transport error names an errno that differs by platform; the
        // news is the same.
        Err(_) => return Ok(Outcome::Unreachable("models.dev could not be reached".into())),
    };
    match res.status().as_u16() {
        304 if !etag.is_empty() => {
            stamp_meta(&path, &etag).map_err(|e| format!("pricing: stamp {}: {e}", meta_path(&path).display()))?;
            Ok(Outcome::Unchanged)
        }
        200 => {
            let new_etag = sanitize_etag(res.headers().get("etag").and_then(|v| v.to_str().ok()).unwrap_or_default());
            let mut body = Vec::new();
            if std::io::Read::read_to_end(&mut std::io::Read::take(res.body_mut().as_reader(), MAX_BODY + 1), &mut body).is_err()
                || body.len() as u64 > MAX_BODY
            {
                return Ok(Outcome::Unreachable("models.dev answered with a body krowk could not read".into()));
            }
            if parse_rates(&body).is_none_or(|t| t.is_empty()) {
                return Ok(Outcome::Unreachable("models.dev answered with something that is not a price file".into()));
            }
            write_atomic(&path, &body).map_err(|e| format!("pricing: write {}: {e}", path.display()))?;
            stamp_meta(&path, if new_etag.is_empty() { &etag } else { &new_etag }).map_err(|e| format!("pricing: stamp {}: {e}", meta_path(&path).display()))?;
            *LOADED.lock().unwrap_or_else(|e| e.into_inner()) = None;
            Ok(Outcome::Refreshed)
        }
        s => Ok(Outcome::Unreachable(format!("models.dev answered HTTP {s}"))),
    }
}

/// Whether the cache file holds prices at all.
fn cache_readable(cache: &Path) -> bool {
    std::fs::read(cache).ok().and_then(|raw| parse_rates(&raw)).is_some_and(|t| !t.is_empty())
}

/// How old the prices are.
#[derive(Debug, Clone, PartialEq)]
pub enum Freshness {
    /// No cache: the snapshot built into krowk prices everything.
    SnapshotOnly,
    /// A cache file that holds no prices: the snapshot is doing the work.
    Unreadable,
    /// A cache, fetched (or confirmed by a 304) at this moment — unknown
    /// when its sidecar is missing or corrupt.
    Cache { fetched_at_ms: Option<i64> },
}

impl Freshness {
    pub fn of(env: &dyn Fn(&str) -> String) -> Freshness {
        let Some(cache) = cache_path(env).filter(|p| p.exists()) else { return Freshness::SnapshotOnly };
        if !cache_readable(&cache) {
            return Freshness::Unreadable;
        }
        Freshness::Cache { fetched_at_ms: fetched_at_ms(&cache) }
    }

    pub fn fetched_at_ms(&self) -> Option<i64> {
        match self {
            Freshness::Cache { fetched_at_ms } => *fetched_at_ms,
            _ => None,
        }
    }

    /// Whole days since the last fetch; none when unknown, or when the fetch
    /// is in the future — a clock that is off, which no age describes.
    pub fn age_days(&self, now_ms: i64) -> Option<i64> {
        self.fetched_at_ms().filter(|f| *f <= now_ms).map(|f| (now_ms - f) / 86_400_000)
    }

    /// "models.dev prices fetched 2026-09-10, 15 days ago", or what answers instead.
    pub fn describe(&self, now_ms: i64) -> String {
        match self {
            Freshness::SnapshotOnly => format!("no price cache — prices come from the models.dev snapshot {SNAPSHOT_DATE} embedded in this build"),
            Freshness::Unreadable => format!("the price cache holds no prices — the models.dev snapshot {SNAPSHOT_DATE} embedded in this build prices everything"),
            Freshness::Cache { fetched_at_ms: None } => "a price cache of unknown age — its fetch time was not recorded".into(),
            Freshness::Cache { fetched_at_ms: Some(ms) } => match self.age_days(now_ms) {
                None => format!("models.dev prices fetched {}, which is in the future — this machine's clock is off", date_of(*ms)),
                Some(0) => format!("models.dev prices fetched {}, today", date_of(*ms)),
                Some(1) => format!("models.dev prices fetched {}, 1 day ago", date_of(*ms)),
                Some(d) => format!("models.dev prices fetched {}, {d} days ago", date_of(*ms)),
            },
        }
    }
}

/// When the cache was last fetched, from its sidecar.
pub fn fetched_at_ms(cache: &Path) -> Option<i64> {
    let meta: Value = serde_json::from_slice(&std::fs::read(meta_path(cache)).ok()?).ok()?;
    meta.get("fetched_at_ms")?.as_i64()
}

fn load_etag(cache: &Path) -> String {
    std::fs::read(meta_path(cache))
        .ok()
        .and_then(|raw| serde_json::from_slice::<Value>(&raw).ok())
        .and_then(|m| m.get("etag")?.as_str().map(sanitize_etag))
        .unwrap_or_default()
}

/// An ETag is visible ASCII and bounded, or it is not replayed.
fn sanitize_etag(etag: &str) -> String {
    if etag.is_empty() || etag.len() > MAX_ETAG || !etag.bytes().all(|b| (0x21..=0x7e).contains(&b)) {
        return String::new();
    }
    etag.to_string()
}

fn stamp_meta(cache: &Path, etag: &str) -> std::io::Result<()> {
    let now = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).map_or(0, |d| d.as_millis() as i64);
    let meta = serde_json::json!({ "etag": etag, "fetched_at_ms": now });
    write_atomic(&meta_path(cache), format!("{meta}\n").as_bytes())
}

fn write_atomic(path: &Path, data: &[u8]) -> std::io::Result<()> {
    let dir = path.parent().unwrap_or(Path::new("."));
    std::fs::create_dir_all(dir)?;
    let tmp = dir.join(format!(".tmp-{}", std::process::id()));
    std::fs::write(&tmp, data)?;
    std::fs::rename(&tmp, path).inspect_err(|_| {
        let _ = std::fs::remove_file(&tmp);
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn both_price_shapes_parse_and_reasoning_falls_back_to_output() {
        let snapshot = parse_rates(EMBEDDED.as_bytes()).unwrap();
        assert!(snapshot.contains_key(&("anthropic".into(), "claude-sonnet-4-6".into())));
        let full = parse_rates(br#"{"p":{"models":{"m":{"cost":{"input":1,"output":2,"reasoning":null}}}}}"#).unwrap();
        let r = full[&("p".into(), "m".into())];
        assert!(!r.has_reasoning);
        let cost = r.cost(Tokens { input: 1_000_000, output: 500_000, reasoning: 500_000, cache_read: -5, ..Tokens::default() });
        assert!((cost - 3.0).abs() < 1e-9);
        assert!(parse_rates(b"not json").is_none());
        let bad_neighbours = parse_rates(br#"{"p":{"models":{"m":{"cost":{"input":1}},"n":null,"o":{"cost":"free"}}}}"#).unwrap();
        assert_eq!(bad_neighbours.len(), 1, "a bad entry costs itself, not its provider");
        assert_eq!(sanitize_etag("W/\"abc\""), "W/\"abc\"");
        assert_eq!(sanitize_etag("bad etag"), "");
    }
}
